1.0 iterate

jakewins opened the pull request:
Two large API changes - one swapping Cursor for Iterator (see commit message for details), and one reorganizing the final API after that change (see commit message for details).

Because this moves quite a few classes in order to simplify onboarding into the API, the commits are quite large. Suggest you read them each on their own, rather than this whole PR in one go.

As far as I know, this now contains all pending API changes for 1.0.

jakewins committed:
Change ResultCursor to Iterator<Result>

When we first designed the result cursor, the intention was to build a
low-overhead API that didn't require allocating lots of objects to move
through the result.

In testing this with developers, there's been a fair bit of negative
feedback and, more concerning - misunderstanding the model. Many
developers are not used to the Cursor pattern, and don't untuitively
imagine it as "moving a view" over the result. This leads to dissonance,
which leads to frustration, which leads to suffering, which leads to the
dark side.

Tipping the scale is a series of experiments with escape analysis we
did. It turns out that escape analysis is powerful enough to avoid heap
allocation even for nested calls like:

    result.next().get("blah").get(0).asInt();

The construct above can be allocated entirely on the stack, even if each
method but the last returns a newly allocated object. This has been
verified in experiments on HotSpot 25.74-b02, Java 8.

Notably, the current implementation does not allow this, since Record
is not allocated directly in the `next` method - but this is of no
significance, because this was the case for the Cursor API as well. The
point is that the API *allows* us to write heap-allocation-free impl in
the future.

This means that we can use an Iterator-based approach, which is MUCH
more intuitive to end users. It also means we can do away with a
significant amount of edge cases ("this method only works if the cursor
is in so-and-so position") and a long slew of methods put in place to
avoid allocation (eg. `single(<fieldName>)` and friends).

Finally, it means there is no longer a distinction between 'retaining'
an individual record and 'viewing' a record. A record is always an
immutable copy, if you store it somewhere it works like any other
object. If you don't, escape analysis will allow allocating it on the
stack.

jakewins committed:
Final restructuring to help discoverability of API.

Move interfaces that are not needed to get going into dedicated
namespaces - summary, util and value - collapse unused interfaces,
and finally remove Identity type.

This creates a root API namespace which should be significantly easier
to get started with, and through which users will discover the value
types and, if they need them, the summary types.

technige commented on driver/src/main/java/org/neo4j/driver/internal/InternalResultStream.java:
```diff
@@ -28,26 +28,25 @@
-import static org.neo4j.driver.v1.Records.recordAsIs;
 
-public class InternalResultCursor extends InternalRecordAccessor implements ResultCursor
+public class InternalResultStream implements ResultStream
```
We should discuss naming. I think we all have different ideas here.

jakewins replied:
Agree, @boggle noted the same thing. I just picked something :) @boggle suggested `RecordResult`. Ideas welcome! Maybe even just `Result`?

jakewins committed:
Final refactoring of API as per meeting this morning

jakewins committed:
Further API polish

- asMap() and asList() now default to List<Object> and
  Map<String,Object>.
- MapValue exposes asMap(), obviating properties()
- Session.close() calls sync, ensuring all outstanding operations
  complete before the session is RESET when it enters the pool.
- StatementRunner, Session, Transaction and StatementResult all
  have lengthy expansions describing the semantics of lazy results

boggle commented:
Looks good.. noticing in passing that Values.asMap() misses a variant that takes a function for mapping over the values

jakewins committed:
Convert valueAsXXX to ofXXX, which fits better with how functions are used.

jakewins commented:
@boggle fixed above. Imma merge this then, since it resolves two outstanding blocking issues.
